packages/modules/http/device.py

```diff
@@ -106,7 +106,7 @@ def read_legacy(argv: List[str]) -> None:
                 "power_path": __extract_url_path(argv[2]),
                 "counter_path": __extract_url_path(argv[3]),
             }
-            num = argv[4]
+            num = int(argv[4])
```

Comment:
Jepp, korrekt. Steht allerdings in Konflikt mit meinem PR #1831 in der ich das Problem auch löse, nur auf eine ganz andere Weise.

Reply:
Ah ok, habe leider die pull requests diesbezüglich heute nicht mehr durchsucht, wollte den fehler bei mir lediglich schnell fixen!
Hab deine Änderung bei mir getestet, leider werden hier die PV Werte nicht aktualisiert und es gibt  leider auch eine exception:
KeyError: 'power_all_path'
    power_all = request_value(self.domain + config["power_all_path"])
  File "/var/www/html/openWB/packages/modules/http/counter.py", line 37, in update
    self._components[component].update()
  File "/var/www/html/openWB/packages/modules/http/device.py", line 65, in update
Traceback (most recent call last):

Wenn du hierzu weitere Infos brauchst melde dich gerne...

Wir können dann die Änderungen gerne zusammenführen, bzw. du noch meine Änderung in wr_http/main.sh übernehmen, da diese noch notwendig ist um das Problem ganz zu lösen.

Reply:
Oh cool, das ist super, dass du das testest, danke! Und gleich auch was gefunden :-). Habe ich gefixt. Wäre toll wenn du jetzt die aktuelle Version nochmal testest.

Reply:
@yankee42 
Deine Bemühungen um korrekte Url's in allen Ehren aber....

Ich behaupte mal das openwb.conf sich nicht von alleine ändert.
Weiterhin angenommen das die User die per SSH darin rumbasteln können wissen was sie tun
Unter diesen beiden Annahmen würde ich die Prüfung genau 1 mal durchführen und zwar bei speichern der Einstellungen.

Also N=1 und nicht 

8640 mal am Tag oder 
259200 mal im Monat oder
3153600 mal im Jahr oder
15768000 mal im 5 Jahren.
Um die Nerven zu schonen könnte man noch ein "Prüflauf/Selbsttest" bei cron{daily|hourly} vorsehen.

Reply:
Hab es nochmal getestet => super, ist gefixt!! Funktioniert jetzt wie erwartet!!

Reply:
@hhoefling 

> Ich behaupte mal das openwb.conf sich nicht von alleine ändert.

Und genau das ist unter anderem ein Grund warum der check gebraucht wird. Leute machen ein Update der openWB und hatten in ihrer alten Konfiguration da unterschiedliche Optionen drin, die jetzt nicht mehr funktionieren. Wo wird das optimalerweise detektiert? Gut, du kannst auch ein Update-Script machen, was das ganze nur einmal macht, aber wo wird das dann angezeigt? Da brauchst du überhaupt erstmal ein System für, dass sowas durch kommt. Das ist wirklich großer Aufwand. Außerdem muss dieses Update-Check dann wieder mit migriert werden, wenn sich was ändert. Da das Upgrade-Skript aber ganz woanders liegt, wird das potentiell vergessen.

> Weiterhin angenommen das die User die per SSH darin rumbasteln können wissen was sie tun

Schlechte Annahme, denn die Konfiguration hier ist alles andere als intuitiv. Dass die Konfiguration so ist, wie sie ist hat historische Gründe. Das System war früher anders und jetzt soll die Konfiguration rückwärtskompatibel sein und das führt zu diesem unintuitiven Konstrukt.

> 15768000 mal im 5 Jahren.

Ähm... Ja und? Hast du Angst dass sich da was abnutzt? Geht es dir um die Laufzeit? Das ganze Skript braucht 2.5 Sekunden, bis Python alle Module initialisiert hat. Das komplette Script was dann noch läuft braucht nur wenige Millisekunden. Die Millisekunden gehen für I/O drauf. Der URL-check hier, da reden wir über Nanosekunden.

Hier mal ein Benchmark. Code:
```python
import re

import pytest


def create_paths_dict(**kwargs):
    regex = re.compile("^(https?://[^/]+)(.*)")
    result = {}
    host_scheme = None
    for key, path in kwargs.items():
        if path == "none":
            result[key] = "none"
        else:
            match = regex.search(path)
            if match is None:
                raise Exception("Invalid URL <" + path + ">: Absolute HTTP or HTTPS URL required")
            if host_scheme is None:
                host_scheme = match.group(1)
            elif host_scheme != match.group(1):
                raise Exception("All URLs must have the same scheme and host. However URLs are: " + str(kwargs))
            result[key] = match.group(2)
    return result


def __extract_url_path(arg):
    return re.search("^(?:https?://[^/]+)?(.*)", arg).group(1)


def old(power_path: str, imported_path: str, exported_path: str, soc_path: str):
    return {
        "power_path": __extract_url_path(power_path),
        "imported_path": __extract_url_path(imported_path),
        "exported_path": __extract_url_path(exported_path),
        "soc_path": __extract_url_path(soc_path),
    }


def new(power_path: str, imported_path: str, exported_path: str, soc_path: str):
    return create_paths_dict(
        power_path=power_path,
        imported_path=imported_path,
        exported_path=exported_path,
        soc_path=soc_path,
    )


@pytest.mark.parametrize("function", [old, new])
def test_benchmark_old(benchmark, function):
    benchmark.pedantic(
        function,
        args=("http://domain/path", "http://domain/path", "http://domain/path", "http://domain/path"),
        iterations=1000,
        rounds=10000
    )
```

```
------------------------------------------------------------------------------------ benchmark: 2 tests ------------------------------------------------------------------------------------
Name (time in us)              Min               Max              Mean            StdDev            Median               IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_benchmark_old[new]     2.0388 (1.0)      6.6800 (1.0)      2.2044 (1.0)      0.1308 (1.05)     2.1842 (1.0)      0.0873 (1.0)       501;353      453.6384 (1.0)       10000        1000
test_benchmark_old[old]     2.2270 (1.09)     8.6198 (1.29)     2.3967 (1.09)     0.1244 (1.0)      2.3826 (1.09)     0.0943 (1.08)      567;303      417.2469 (0.92)      10000        1000
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
==================================================================================== 2 passed in 46.10s =====================================================================================
```
Die Aussagekraft von so einem Benachmark ist zwar auch immer so eine Sache, aber mit dem Test komme ich immerhin schonmal dahin, dass der neue Code sogar 9% SCHNELLER (!) ist, als der alte. Das liegt natürlich nicht daran, dass da ein Check dazu gekommen ist, sondern, dass der Code generell anders aufgebaut ist (tatsächlich fällt mir nichtmal ein Grund ein, warum das so ist, ich hätte schon erwartet, dass der alte Code schneller ist, aber ich habe das Benchmark mehrmals laufen lassen und es ist immer ungefähr gleich). Es zeigt wie dermaßen überhaupt garnicht dieser Check ins Gewicht fällt.

Hinzu kommt noch, dass das ständige Ausführen ein reines openWB 1 Problem ist. In openWB 2 wird der check genau einmal beim Initialisieren gemacht.

Es gibt schon ein paar Performance-Flaschenhälse hier in oWB. Insbesondere das Hochfahren der Pythonumgebung ist so ein Fall (habe ich übrigens mir auch mal angesehen. 2.5 Sekunden Braucht das Teil. 25% der Zeit geht auf das Initialisieren von der `requests`-Bibliothek, weitere 25% der Zeit auf `paho-mqtt`. Zwei Bibliotheken die wirklich zentral sind und man auch schlecht entfernen/ersetzen kann. Der Ansatz die Scripte nur einmal zu starten und am Laufen zu halten würde jedenfalls Milliardenfach mehr ausmachen als hier sich über den Kram zu unterhalten.

> Um die Nerven zu schonen könnte man noch ein "Prüflauf/Selbsttest" bei cron{daily|hourly} vorsehen.

Die zusätzliche Startzeit für ein weiteres Skript könnte den Performancevorteil bereits wieder zunichte machen. Dafür ist es deutlich weniger nützlich. Denn dass das dann beim Debuggen gesehen wird muss auch sichergestellt werden. Wenn man morgens die Konfig ändert (oder ein Update macht) und am nächsten Tag findet man einen Eintrag von einem Prüflauf im log, dann ist schon viel Zeit vergeudet worden. Außerdem denkt natülich garantiert niemand daran, wenn sich am Script was ändert auch den cron anzupassen.

Die Prüfung der URLs dagegen ist sinnvoll. Es ist ein minimalster Aufwand und es erspart dem Nutzer dem aus irgend einem Grund unterschiedliche Domains in die Konfig gespült kommen viel Kopfzerbrechen.

Reply:
@matzempc Vielen Dank für das Testen!

Reply:
> Gut, du kannst auch ein Update-Script machen, was das ganze nur einmal macht, aber wo wird das dann angezeigt?  
>
Ein Update Script soll eine funktionierende Konfiguration in einen neue funktionierenden Konfiguration überführen und idealerweise einen Bericht darüber anzeigen. Z.B einen Logdatei die automatisch nach eventuellem Neustart angezeigt wird. Deine Debug Meldungen vergammel ja auch im debug.log herum.
Wenn das Update das nicht kann gehört es nicht freigegeben.


>Das ganze Skript braucht 2.5 Sekunden, bis Python alle Module initialisiert hat. 

Wenn das wirklich stimmt kanst du die 1.9x sowieso komplett vergessen.
dann kannst du nur 3 maxumal 4 Untermodule als Pyhton aufrufen, pro 10 Sekunden Regelschleife.

bezug 1x
bat 1x
WR 1x
SOC via http 1x
Tarifinfo 1x
2 Shelly 2x
2'ter LP 1x
Leistungsdoule dafür 1x

also 9 mals Phyton Startup time von 2.5 Sekunden  :-( Leider ein Nogo

Die ganzen 1.9'er Module Scripte dürfem IN SUMME auf einem PI-3B+ nicht über 7-8 Sekunden gehen.